check.go

```diff
@@ -971,6 +1011,30 @@ func (c *checkDirectory) header(hdr *tar.Header) {
 			c.remove(path.Join(dir, base[len(archive.WhiteoutPrefix):]))
 		}
 	} else {
+		if hdr.Typeflag == tar.TypeLink {
+			// look up the attributes of the target of the hard link
+			// n.b. by convention, Linkname is always relative to the
+			// root directory of the archive, which is not always the
+			// same as being relative to hdr.Name
+			directory := c
+			for _, component := range strings.Split(hdr.Linkname, "/") {
```

Comment:
This seems incorrect for
- targets that end with `/.` (the attributes are not set)
- targets with with `".."` elements, I think.
- targets that link to a “subdirectory within a file”, e.g. for an existing non-directory `a/b` and a link target `a/b/c` (incorrectly uses attributes of the last matching element)

Layers with the “subdirectory within a file” cases should be impossible to extract, so I’m not sure how much we care. If a subdirectory were corrupted on disk and replaced with a file, that should be detected when checking the link target’s tar header, shouldn’t it?

A `filepath.Clean` here would be consistent with how `pkg/archive.createTarFile` implicitly does one via `filepath.Join`.

Reply:
Adding a `path.Clean()` here, which should strip any final "/."s off of the `Linkname` value.
If a subdirectory were replaced with a file in a layer, we'd notice the difference due to its type being different from what we expected, if nothing else.
We should skip ".." components, too, because `path.Clean()` should be taking out any such components that aren't at the beginning of the path, but since it's a relative path, it isn't going to remove them if the path starts with "../".